Asynchronous loading of tables

serxa opened the pull request:
Should be merged after https://github.com/ClickHouse/ClickHouse/pull/48923
Fixes https://github.com/ClickHouse/ClickHouse/issues/43424
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added server setting `async_load_databases` for asynchronous loading of databases and tables. Speeds up the server start time. Applies to databases with Ordinary, Atomic and Replicated engines. Their tables load metadata asynchronously. Query to a table increases the priority of the load job and waits for it to be done. Added table `system.async_loader`.

### Documentation entry for user-facing changes

- [X] Documentation is written (mandatory for new features)

serxa committed:
wip AsyncLoader

serxa committed:
make t work, add simple test

serxa committed:
fix style

serxa committed:
add dependency cycle detection

serxa committed:
add simple randomized test

serxa committed:
add tests for job status/wait/cancel

serxa committed:
add more tests for cancel

serxa committed:
fix remove of executing tasks + add test

serxa committed:
allow cancels on RandomTasks test shutdown

serxa committed:
add concurrency test

serxa committed:
add  overload test

serxa committed:
test deps are completed + refactoring

serxa committed:
add LoadStatus::CANCELED

serxa committed:
fix scheduling of jobs with dependencies on finished jobs

serxa committed:
add ScheduleJobWithCanceledDependencies test

serxa committed:
fix worker shutdown

serxa committed:
add random test with dependent tasks

serxa committed:
make performance-inefficient-vector-operation,-warnings-as-errors happy

serxa committed:
add test for static priorities

serxa committed:
fix test

serxa committed:
allow to change max number of workers in runtime

serxa committed:
add a good comment

serxa committed:
fix

serxa committed:
Update src/Common/AsyncLoader.h

serxa committed:
Update src/Common/AsyncLoader.h

serxa committed:
split AsyncLoader into .h and .cpp files + more docs

serxa committed:
fix comment

serxa committed:
fix double finish() during cancellation of a job with multiple paths

serxa committed:
add test for dynamic prioritization

serxa committed:
fix tidy build

serxa committed:
fix data race in test

serxa committed:
fix typos

serxa committed:
fix more comments

serxa committed:
better job shutdown

serxa committed:
typo

serxa committed:
Update src/Common/AsyncLoader.h

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

serxa committed:
review fixes

serxa committed:
Update src/Common/AsyncLoader.h

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

serxa committed:
improve AsyncLoader interface

serxa committed:
wip async tables loading

serxa committed:
more work

serxa committed:
fix build

serxa committed:
fix tidy build

serxa committed:
fix tidy build

serxa committed:
log progress in AsyncLoader

serxa committed:
working on async loading in TablesLoader

serxa committed:
schedule() jobs recursively with its dependencies

serxa committed:
more helpers for AsyncLoader

serxa committed:
integrate with Server and TablesLoader

serxa committed:
Merge branch 'master' into async-loader-integration

robot-clickhouse commented:
<!-- automatic status comment for PR #49351 from ClickHouse/ClickHouse:async-loader-integration -->
This is an automated comment for commit 8dab49f328dd13072055099589e3cddfeb5d3114 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49351/8dab49f328dd13072055099589e3cddfeb5d3114/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

serxa committed:
fix typo

serxa committed:
cleanup

serxa committed:
improve helper functions

serxa committed:
implement async loading interface in database engines

serxa committed:
Update src/Common/AsyncLoader.cpp

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

serxa committed:
Update src/Common/AsyncLoader.h

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

serxa committed:
Update src/Common/AsyncLoader.h

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

serxa committed:
Update src/Common/AsyncLoader.h

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

serxa committed:
add AsyncLoader introspection

serxa committed:
add system.async_loader table

serxa committed:
do not pollute unittests output with progress

serxa committed:
fix style

serxa committed:
fix style, get rid of DatabaseOrdinaryThreads* metrics

serxa commented:
At this point, we should have everything refactored from sync into async loading with sync wait. So there should be no visible difference whatsoever. All tests should work. The next step is to remove sync-wait points before server startup for tables that should load in an async way. And add load job waits to all the queries that require such a table to be loaded.

But first, we should check that all tests are green, and fix them if not.

serxa commented:
Fast test failed with:
```
2023.05.04 05:30:20.515372 [ 154620 ] {} <Information> Application: Loading metadata from /fasttest-workspace/db-fasttest/
2023.05.04 05:30:20.516273 [ 154667 ] {} <Debug> DNSResolver: Updating DNS cache
2023.05.04 05:30:20.517421 [ 154667 ] {} <Debug> DNSResolver: Updated DNS cache
2023.05.04 05:30:20.517726 [ 154620 ] {} <Information> DatabaseAtomic (system): Metadata processed, database system has 0 tables and 0 dictionaries in total.
2023.05.04 05:30:20.517739 [ 154620 ] {} <Information> TablesLoader: Parsed metadata of 0 tables in 1 databases in 6.72e-05 sec
2023.05.04 05:30:20.517746 [ 154620 ] {} <Trace> ReferentialDeps: No tables
2023.05.04 05:30:20.517751 [ 154620 ] {} <Trace> LoadingDeps: No tables
2023.05.04 05:30:20.525828 [ 154620 ] {} <Error> Application: Caught exception while loading metadata: Code: 82. DB::Exception: Database information_schema already exists.: while loading database `information_schema` from path /fasttest-workspace/db-fasttest/metadata/information_schema. (DATABASE_ALREADY_EXISTS), Stack trace (when copying this message, always include the lines below):

0. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/exception:134: Poco::Exception::Exception(String const&, int) @ 0x12f816b3 in /fasttest-workspace/build/programs/clickhouse
1. /fasttest-workspace/build/./src/Common/Exception.cpp:91: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0xcb65d95 in /fasttest-workspace/build/programs/clickhouse
2. DB::Exception::Exception<String&>(int, FormatStringHelperImpl<std::type_identity<String&>::type>, String&) @ 0x84c8344 in /fasttest-workspace/build/programs/clickhouse
3. /fasttest-workspace/build/./src/Interpreters/InterpreterCreateQuery.cpp:128: DB::InterpreterCreateQuery::createDatabase(DB::ASTCreateQuery&) @ 0x110d57e0 in /fasttest-workspace/build/programs/clickhouse
4. /fasttest-workspace/build/./src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::execute() @ 0x110eb7cd in /fasttest-workspace/build/programs/clickhouse
5. /fasttest-workspace/build/./src/Interpreters/loadMetadata.cpp:0: DB::executeCreateQuery(String const&, std::shared_ptr<DB::Context>, String const&, String const&, bool, bool) @ 0x11b4a958 in /fasttest-workspace/build/programs/clickhouse
6. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/string:1499: DB::loadDatabase(std::shared_ptr<DB::Context>, String const&, String const&, bool) @ 0x11b45085 in /fasttest-workspace/build/programs/clickhouse
7. /fasttest-workspace/build/./src/Interpreters/loadMetadata.cpp:0: DB::loadSystemDatabaseImpl(std::shared_ptr<DB::Context>, String const&, String const&) @ 0x11b4a583 in /fasttest-workspace/build/programs/clickhouse
8. /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/string:1499: DB::loadMetadataSystem(std::shared_ptr<DB::Context>) @ 0x11b49b12 in /fasttest-workspace/build/programs/clickhouse
9. /fasttest-workspace/build/./programs/server/Server.cpp:0: DB::Server::main(std::vector<String, std::allocator<String>> const&) @ 0xcbed26f in /fasttest-workspace/build/programs/clickhouse
10. /fasttest-workspace/build/./base/poco/Util/src/Application.cpp:0: Poco::Util::Application::run() @ 0x12ca9b86 in /fasttest-workspace/build/programs/clickhouse
11. /fasttest-workspace/build/./programs/server/Server.cpp:476: DB::Server::run() @ 0xcbe1410 in /fasttest-workspace/build/programs/clickhouse
12. /fasttest-workspace/build/./base/poco/Util/src/ServerApplication.cpp:132: Poco::Util::ServerApplication::run(int, char**) @ 0x12cbdc32 in /fasttest-workspace/build/programs/clickhouse
13. /fasttest-workspace/build/./programs/server/Server.cpp:0: mainEntryClickHouseServer(int, char**) @ 0xcbde9bc in /fasttest-workspace/build/programs/clickhouse
14. /fasttest-workspace/build/./programs/main.cpp:0: main @ 0x8146171 in /fasttest-workspace/build/programs/clickhouse
15. __libc_start_main @ 0x7f97b68a8083 in ?
16. _start @ 0x814562e in /fasttest-workspace/build/programs/clickhouse
 (version 23.5.1.1)
```

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
start loader, fix helpers, fix deadlock

serxa commented:
Previous problem fixed, now server start but after few tests SEGFAULTs on CREATE TABLE query:
```
2023.05.10 23:14:47.350665 [ 246209 ] {} <Fatal> BaseDaemon: (version 23.5.1.1, build id: 0612A6F35AB2A48D3EA0DE088AE7BB993212450B) (from thread 200968) (query_id: c6770196-b0e5-48a9-97bf-60fe9093f3f8) (query: CREATE TABLE test_l1yoto9k_db.tab_rmt_ok (x UInt32) engine = ReplicatedMergeTree order by x;) Received signal Segmentation fault (11)
2023.05.10 23:14:47.350683 [ 246209 ] {} <Fatal> BaseDaemon: Address: 0x4f. Access: read. Address not mapped to object.
2023.05.10 23:14:47.350699 [ 246209 ] {} <Fatal> BaseDaemon: Stack trace: 0x0000000010f22e01 0x0000000011171bd9 0x000000001117a681 0x0000000011bb2638 0x0000000011bafa96 0x000000001269738b 0x00000000126a6019 0x0000000013008d07 0x00000000130091ed 0x000000001306e347 0x000000001306bf03 0x00007fa538db1609 0x00007fa53903a133
2023.05.10 23:14:47.412423 [ 246209 ] {} <Fatal> BaseDaemon: 3.1. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/string:1499: String::__is_long[abi:v15000]() const
2023.05.10 23:14:47.412448 [ 246209 ] {} <Fatal> BaseDaemon: 3.2. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/string:2012: basic_string
2023.05.10 23:14:47.412466 [ 246209 ] {} <Fatal> BaseDaemon: 3.3. inlined from /fasttest-workspace/build/./src/Interpreters/DDLWorker.h:62: DB::DDLWorker::getCommonHostID() const
2023.05.10 23:14:47.412483 [ 246209 ] {} <Fatal> BaseDaemon: 3. /fasttest-workspace/build/./src/Databases/DatabaseReplicated.cpp:703: DB::DatabaseReplicated::tryEnqueueReplicatedDDL(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, bool) @ 0x0000000010f22e01 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.484263 [ 246209 ] {} <Fatal> BaseDaemon: 4.1. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<DB::DDLGuard>::operator()[abi:v15000](DB::DDLGuard*) const
2023.05.10 23:14:47.591603 [ 246209 ] {} <Fatal> BaseDaemon: 4.2. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<DB::DDLGuard, std::default_delete<DB::DDLGuard>>::reset[abi:v15000](DB::DDLGuard*)
2023.05.10 23:14:47.592190 [ 246209 ] {} <Fatal> BaseDaemon: 4.3. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2023.05.10 23:14:47.592354 [ 246209 ] {} <Fatal> BaseDaemon: 4. /fasttest-workspace/build/./src/Interpreters/InterpreterCreateQuery.cpp:1267: DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x0000000011171bd9 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.667562 [ 246209 ] {} <Fatal> BaseDaemon: 5. /fasttest-workspace/build/./src/Interpreters/InterpreterCreateQuery.cpp:1716: DB::InterpreterCreateQuery::execute() @ 0x000000001117a681 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.724520 [ 246209 ] {} <Fatal> BaseDaemon: 6. /fasttest-workspace/build/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000011bb2638 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.774675 [ 246209 ] {} <Fatal> BaseDaemon: 7. /fasttest-workspace/build/./src/Interpreters/executeQuery.cpp:1173: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x0000000011bafa96 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.825444 [ 246209 ] {} <Fatal> BaseDaemon: 8. /fasttest-workspace/build/./src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x000000001269738b in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.880884 [ 246209 ] {} <Fatal> BaseDaemon: 9. /fasttest-workspace/build/./src/Server/TCPHandler.cpp:2045: DB::TCPHandler::run() @ 0x00000000126a6019 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.882546 [ 246209 ] {} <Fatal> BaseDaemon: 10. /fasttest-workspace/build/./base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x0000000013008d07 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.885230 [ 246209 ] {} <Fatal> BaseDaemon: 11.1. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<Poco::Net::TCPServerConnection>::operator()[abi:v15000](Poco::Net::TCPServerConnection*) const
2023.05.10 23:14:47.885251 [ 246209 ] {} <Fatal> BaseDaemon: 11.2. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<Poco::Net::TCPServerConnection, std::default_delete<Poco::Net::TCPServerConnection>>::reset[abi:v15000](Poco::Net::TCPServerConnection*)
2023.05.10 23:14:47.885271 [ 246209 ] {} <Fatal> BaseDaemon: 11.3. inlined from /fasttest-workspace/build/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2023.05.10 23:14:47.885280 [ 246209 ] {} <Fatal> BaseDaemon: 11. /fasttest-workspace/build/./base/poco/Net/src/TCPServerDispatcher.cpp:116: Poco::Net::TCPServerDispatcher::run() @ 0x00000000130091ed in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.888398 [ 246209 ] {} <Fatal> BaseDaemon: 12. /fasttest-workspace/build/./base/poco/Foundation/src/ThreadPool.cpp:202: Poco::PooledThread::run() @ 0x000000001306e347 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.891220 [ 246209 ] {} <Fatal> BaseDaemon: 13.1. inlined from /fasttest-workspace/build/./base/poco/Foundation/include/Poco/SharedPtr.h:139: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::assign(Poco::Runnable*)
2023.05.10 23:14:47.891241 [ 246209 ] {} <Fatal> BaseDaemon: 13.2. inlined from /fasttest-workspace/build/./base/poco/Foundation/include/Poco/SharedPtr.h:180: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::operator=(Poco::Runnable*)
2023.05.10 23:14:47.891262 [ 246209 ] {} <Fatal> BaseDaemon: 13. /fasttest-workspace/build/./base/poco/Foundation/src/Thread_POSIX.cpp:350: Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001306bf03 in /fasttest-workspace/build/programs/clickhouse
2023.05.10 23:14:47.891272 [ 246209 ] {} <Fatal> BaseDaemon: 14. ? @ 0x00007fa538db1609 in ?
2023.05.10 23:14:47.891305 [ 246209 ] {} <Fatal> BaseDaemon: 15. __clone @ 0x00007fa53903a133 in ?
```

serxa committed:
fix startup of database w/o tables

serxa committed:
fix LocalServer fast-tests

serxa committed:
temporarily disable JobFailure test

serxa committed:
fix disable test

serxa committed:
add async_load_databases server setting

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix

serxa committed:
fix metrics docs

serxa committed:
refact

serxa committed:
fix

serxa committed:
fix style

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix conflicts after merge

serxa commented:
There are missing tables, looks like not all tables are completely loaded during start at the point:
```
2023-05-14 16:50:24 Code: 60. DB::Exception: Table `db_ordinary`.`log_table` doesn't exist. (UNKNOWN_TABLE)
```

serxa committed:
fix `UNKNOWN_TABLE` problem during async startup

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix style

serxa committed:
refactoring

serxa committed:
fix tidy build

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
rework async loader initialization and reconfiguration for bg/fg pools

serxa committed:
use pool ids instead of priorities in load jobs

serxa committed:
add helpers for prioritize+wait calls and fix build

serxa committed:
fix `system.async_loader`

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix server

serxa committed:
fix clickhouse-local

serxa committed:
Merge branch 'master' into async-loader-integration

serxa commented:
`01160_table_dependencies` failed with timeout - looks related.

serxa committed:
wait db load before DROP table from Atomic Database

serxa committed:
clean

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'async-loader-integration' of github.com:ClickHouse/ClickHouse into async-loader-integration

serxa committed:
add `utils/async_loader_graph` graph rendering tool

serxa committed:
fix usage

serxa committed:
introduce unique job_id in async_loader

serxa committed:
reply on job_id instead of job name

serxa committed:
fix style

serxa committed:
add async_laoder deadlock detection on job waiting

serxa committed:
fix style

serxa committed:
spell

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa commented:
I added a visualizing tool for `system.async_loader` to be able to diagnose issues with async loading and tested it with currently failing test, here is the [output](https://pastila.nl/?034a4092/3dc911b60c75de7291492d5345c21ded.html). I also added deadlock detection for 3 types of deadlock possible due to wrong usage of AsyncLoader. Now it throws LOGICAL_ERROR instead of hanging:
```
Logical error: 'Load job 'load table test_mqxz9x3y.join' waits for dependent load job 'startup table test_mqxz9x3y.dict1''
```

serxa committed:
fix

serxa committed:
do not use table uuid in TablesLoader

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
typos

serxa committed:
add deadlock resolution of "blocked pool"

serxa committed:
unlimit queues and threads in `ThreadPool`

serxa committed:
more testing for suspendend workers

serxa committed:
fix reverse locking problem

serxa committed:
add test for subjobs

serxa committed:
Merge branch 'master' into async-loader-integration

serxa commented:
Upgrade check (msan): 
```
Logical error: \'Load job \'load table test_01676.table\' (priority=1) waits for lower-priority load job \'startup table test_01676.dict\' (priority=2)\'.
Logical error: \'Load job \'load table 02097_db.test_table_default\' (priority=1) waits for lower-priority load job \'startup table 02097_db.test_dictionary\' (priority=2)\'.
Logical error: \'Load job \'load table test_3.dep\' (priority=1) waits for lower-priority load job \'startup table test_3.id_join\' (priority=2)\'.
```

backtrace: https://pastila.nl/?00bfc54a/99d538cebe13ead519318cb8d61d3409

serxa committed:
fix "wait for not scheduled" deadlock in `loadMetadata()`

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
use current priority for waits from load jobs

serxa committed:
Merge branch 'master' into async-loader-integration

serxa commented:
`test_dictionaries_dependency_xml` has two types of "waits for lower-priority load job" errors:
1. `Load job 'load table default.src' failed: Code: 49. DB::Exception: Load job 'load table default.src' (priority=1) waits for lower-priority load job 'startup table system.join' (priority=2): While processing _CAST(joinGet('system.join', 'm', CAST('1', 'int')) AS m_tmp_alter7136918295077541758, 'Int32') AS m: default expression and column type are incompatible.: Cannot attach table `default`.`src` from metadata file /var/lib/clickhouse/store/7e9/7e9ca720-f3e5-44db-8026-12f788d97688/src.sql from query ATTACH TABLE default.src UUID '5a88f59a-9985-41d9-9e82-5cf476acb875' (`n` Int32, `m` Int32 DEFAULT joinGet('system.join', 'm', CAST('1', 'int')), `t` Nullable(Int32) DEFAULT dictGetOrNull('a.d', 'm', toUInt64(3)), `k` Int32 DEFAULT dictGet('a.d', 'm', toUInt64(4))) ENGINE = MergeTree ORDER BY n SETTINGS index_granularity = 8192.` [backtrace](https://pastila.nl/?003a868e/0e8d23230e15372a45131f757ebe3100)
2. `Load job 'load table default.src' (priority=1) waits for lower-priority load job 'startup table syst
em.join' (priority=2): While processing _CAST(joinGet('system.join', 'm', CAST('1', 'int')) AS m_tmp_alter7136918295077541758, 'Int32') AS m: default expression and column type are incompatible.: Cannot attach
 table `default`.`src` from metadata file /var/lib/clickhouse/store/7e9/7e9ca720-f3e5-44db-8026-12f788d97688/src.sql from query ATTACH TABLE default.src UUID '5a88f59a-9985-41d9-9e82-5cf476acb875' (`n` Int32, 
`m` Int32 DEFAULT joinGet('system.join', 'm', CAST('1', 'int')), `t` Nullable(Int32) DEFAULT dictGetOrNull('a.d', 'm', toUInt64(3)), `k` Int32 DEFAULT dictGet('a.d', 'm', toUInt64(4))) ENGINE = MergeTree ORDER
 BY n SETTINGS index_granularity = 8192.` [backtrace](https://pastila.nl/?003a868e/f5bfc014c48a0b2382b595f9ae383eb8)

serxa committed:
fix prioritization/schedule race

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
add auto-resolution for all resolvable deadlock types

serxa commented:
In `test_dictionaries_dependency_xml` test table `default.src` depends on `system.join` table from `system` database. During server start, we first load all system tables, but not start'em up, then load all other databases. So wait added to `getTable()` and used due to mentioned dependency led to "wait not scheduled job" deadlock type.

I added auto-resolution for this deadlock type by implicit scheduling of job inside `wait()`. This seems to be controversial, but we already have similar logic in `schedule()`: it implicitly schedules all the dependencies to avoid deadlocks. From this perspective, implicit scheduling of the job inside `wait()` looks like the next logical step in automatic deadlock resolution.

To make this resolution possible I changed the AsyncLoader API. Method `wait()` of a LoadJob has been moved to AsyncLoader. So we now have all control function there: `AsyncLoader::prioritize(job, pool)`,  `AsyncLoader::schedule(job)` and `AsyncLoader::wait(job)`.

Automatic resolution of "priority inversion" deadlock has also been added instead of throwing LOGICAL_ERROR. This became possible after a rewrite of `wait()` function. It is implemented using priority inheritance, i.e. implicit call to `prioritize()` inside `wait()` iff deadlock is detected.

So the only type of deadlock that is not resolved automatically is "wait dependent" deadlock (when job A function waits for job B that depends on job A). IMO It is not possible to resolve it without rewriting the actual jobs code.

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
avoid using bg pool during non-async start

serxa committed:
remove debug print

serxa committed:
cleanup

serxa committed:
fix

serxa commented:
Test `01193_metadata_loading` is now failing. It checks that `ATTACH DATABASE` query is fast enough. I checked that this kind of queries became significantly slower:
```
master - [124,120,120,119,165] ms
    PR - [750,818,806,726,819] ms
```
Investigating...

serxa committed:
fix: schedule all the jobs before waiting one-by-one

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix build

serxa committed:
add simple integrational tests

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

robot-clickhouse committed:
Automatic style fix

serxa committed:
fix style

serxa committed:
Merge branch 'async-loader-integration' of github.com:ClickHouse/ClickHouse into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa commented on src/Storages/System/StorageSystemAsyncLoader.h:
```diff
@@ -0,0 +1,26 @@
+class StorageSystemAsyncLoader final : public IStorageSystemOneBlock<StorageSystemAsyncLoader>
+{
+public:
+    std::string getName() const override { return "SystemReplicatedFetches"; }
```
```suggestion
    std::string getName() const override { return "SystemAsyncLoader"; }
```

serxa committed:
Update src/Storages/System/StorageSystemAsyncLoader.h

tavplubix commented on programs/server/Server.cpp:
```diff
@@ -1343,6 +1344,12 @@ try
 
+            auto fg_pool_size = server_settings_.async_loader_foreground_pool_size;
+            auto bg_pool_size = server_settings_.async_loader_background_pool_size;
+            global_context->getAsyncLoader().setMaxThreads(AsyncLoaderPoolId::Foreground, fg_pool_size ? fg_pool_size : getNumberOfPhysicalCPUCores());
+            global_context->getAsyncLoader().setMaxThreads(AsyncLoaderPoolId::BackgroundLoad, bg_pool_size ? bg_pool_size : getNumberOfPhysicalCPUCores());
+            global_context->getAsyncLoader().setMaxThreads(AsyncLoaderPoolId::BackgroundStartup, bg_pool_size ? bg_pool_size : getNumberOfPhysicalCPUCores());
```
Maybe we can make `setMaxThreads` accept 0 (but it's not necessary)

tavplubix commented on src/Common/AsyncLoader.cpp:
```diff
@@ -179,9 +168,9 @@ AsyncLoader::AsyncLoader(std::vector<PoolInitializer> pool_initializers, bool lo
-                init.max_threads,
-                /* max_free_threads = */ 0,
-                init.max_threads),
+                /* max_threads = */ std::numeric_limits<size_t>::max(), // Unlimited number of threads, we do worker management ourselves
```
Should we pass some safety constraint just in case?

serxa replied:
I think it is not necessary. And it is hard to pick a value.

tavplubix replied:
> And it is hard to pick a value

Is it becasue of [this](https://github.com/ClickHouse/ClickHouse/pull/49351#discussion_r1317869814)?

serxa replied:
Yes, exactly due to possible suspended workers

tavplubix commented on src/Common/AsyncLoader.cpp:
```diff
@@ -227,27 +216,28 @@ void AsyncLoader::stop()
-void AsyncLoader::schedule(const std::vector<LoadTaskPtr> & tasks)
+void AsyncLoader::schedule(const LoadTaskPtrs & tasks)
 {
+    // TODO(serxa): optimize it to avoid creating new all_jobs - looks like unnecessary allocations
     LoadJobSet all_jobs;
```
I don't think it's a problem for current usecase

tavplubix commented on src/Databases/DatabaseOrdinary.cpp:
```diff
@@ -231,54 +130,128 @@ void DatabaseOrdinary::loadTablesMetadata(ContextPtr local_context, ParsedTables
+    return makeLoadTask(async_loader, {job});
+}
+
+// TODO(serxa): implement
+// DatabaseTablesIteratorPtr DatabaseOrdinary::getTablesIterator(ContextPtr local_context, const DatabaseOnDisk::FilterByNameFunction & filter_by_table_name) const
+// }
```
Why do the tests pass without waiting for tables to be loaded in tables iterator?

serxa replied:
Completely forget about it. Should implement it. Tests are too simple.

serxa replied:
Fixed

tavplubix commented on src/Databases/DatabaseOrdinary.cpp:
```diff
@@ -231,54 +130,128 @@ void DatabaseOrdinary::loadTablesMetadata(ContextPtr local_context, ParsedTables
+// DatabaseTablesIteratorPtr DatabaseOrdinary::getTablesIterator(ContextPtr local_context, const DatabaseOnDisk::FilterByNameFunction & filter_by_table_name) const
+// }
+
+StoragePtr DatabaseOrdinary::tryGetTable(const String & name, ContextPtr local_context) const
```
Do we also need to wait in `createTable`/`dropTable`/`detachTablePermanently`/`renameTable`/`alterTable`/etc?

serxa replied:
Yes, done

tavplubix commented on src/Databases/DatabaseAtomic.cpp:
```diff
@@ -112,6 +113,9 @@ StoragePtr DatabaseAtomic::detachTable(ContextPtr /* context */, const String &
 
 void DatabaseAtomic::dropTable(ContextPtr local_context, const String & table_name, bool sync)
 {
+    // To DROP tables we need the database to be started up (including all the tables)
+    waitLoad(currentPoolOr(AsyncLoaderPoolId::Foreground), getStartupTask());
```
What about other DDLs, such as `DETACH TABLE PERMANENTLY`?

tavplubix commented on src/Databases/TablesLoader.cpp:
```diff
@@ -76,17 +72,66 @@ void TablesLoader::loadTables()
 {
-    /// Startup tables after all tables are loaded. Background tasks (merges, mutations, etc) may slow down data parts loading.
-    for (auto & database : databases)
-        database.second->startupTables(pool, strictness_mode);
-}
+    LoadTaskPtrs result;
+    std::unordered_map<String, LoadTaskPtrs> startup_database; /// database name -> all its tables startup tasks
+
+    for (const auto & table_id : all_loading_dependencies.getTablesSortedByDependency())
```
We can startup tables in any order, `getTablesSortedByDependency` is not necessary

tavplubix commented on src/Interpreters/loadMetadata.cpp:
```diff
@@ -230,19 +235,33 @@ void loadMetadata(ContextMutablePtr context, const String & default_database_nam
-        }
-        catch (...)
-        {
-            tryLogCurrentException("Load metadata", "Can't remove force restore file to enable data sanity checks");
-        }
+        LOG_INFO(log, "Start synchronous loading of databases");
+
+        // Note that wait implicitly calls schedule
+        waitLoad(AsyncLoaderPoolId::Foreground, load_tasks); // First prioritize, schedule and wait all the load table tasks
+        waitLoad(AsyncLoaderPoolId::Foreground, startup_tasks); // Only then prioritize, schedule and wait all the startup tasks
+        return {};
+    }
+    else
```
Please replace `if not ... else ...` with `if .. else ...`

tavplubix commented on src/Interpreters/InterpreterCreateQuery.cpp:
```diff
@@ -321,11 +322,10 @@ BlockIO InterpreterCreateQuery::createDatabase(ASTCreateQuery & create)
             TablesLoader loader{getContext()->getGlobalContext(), {{database_name, database}}, mode};
-            loader.loadTables();
-            loader.startupTables();
+            waitLoad(currentPoolOr(AsyncLoaderPoolId::Foreground), loader.loadTablesAsync());
+            waitLoad(currentPoolOr(AsyncLoaderPoolId::Foreground), loader.startupTablesAsync());
```
Let's support async loading here. It will be useful for tests: we will be able to test async loading using DETACH/ATTACH DATABASE without restarting the server. So it will be possible to write functional tests instead of integration tests

serxa replied:
Done. What kind of functional test could you suggest with this feature? What interesting things can we do in between attach/detach?

serxa commented on docs/en/operations/server-configuration-parameters/settings.md:
```diff
@@ -1628,6 +1628,45 @@ Default value: `0.5`.
+-   Any positive integer.
+-   Zero. Use all available CPUs.
+
+Default value: 16.
```
Yeah, 0 looks better, everybody could lower it if he wants to leave space for query execution while background loading.

serxa replied:
Yeah, 0 looks better, everybody could lower it if he wants to leave space for query execution while background loading.

tavplubix replied:
Also, we can use half of available CPUs for 0 (so it will leave space for query execution by default, and will dynamically choose a reasonable number of threads in current env)

serxa replied:
Now I'm not sure it is a good idea because with such a default, the server will start longer (I mean load all tables; port will open faster of course) after enabling the feature.

serxa replied:
We discussed it and decided to leave "0 means all the cpus" and leave OS to deal with the excess number of threads. Default fixed to be zero.

serxa commented on docs/en/operations/server-configuration-parameters/settings.md:
````diff
@@ -1628,6 +1628,45 @@ Default value: `0.5`.
+<async_load_databases>true</async_load_databases>
+```
+
+## async_loader_foreground_pool_size {#async_loader_foreground_pool_size}
````
I wanted to keep a common prefix, but you are right "foreground" + "async" sounds not so good. Actually, this pool is used by (1) sync loading tables at server start. But we also (2) switch jobs from the background pool to this pool when there is a query waiting on the table. Not sure that "sync" reflects this. Maybe just remove "async"?
 - `foreground_tables_loader_pool` and `background_tables_loader_pool`
 - or just `tables_loader_pool` and `background_tables_loader_pool`

serxa replied:
I wanted to keep a common prefix, but you are right "foreground" + "async" sounds not so good. Actually, this pool is used by (1) sync loading tables at server start. But we also (2) switch jobs from the background pool to this pool when there is a query waiting on the table. Not sure that "sync" reflects this. Maybe just remove "async"?
 - `foreground_tables_loader_pool` and `background_tables_loader_pool`
 - or just `tables_loader_pool` and `background_tables_loader_pool`

tavplubix replied:
Yes, sounds good

serxa replied:
Done

serxa commented on tests/integration/test_async_load_databases/test.py:
Sounds reasonable. Given me a small hint please on how to enable it.

tavplubix replied:
Stress: https://github.com/ClickHouse/ClickHouse/blob/3acb10b8bc08077c9ac39149517709ddc2c3163a/tests/ci/stress.py#L12 or https://github.com/ClickHouse/ClickHouse/blob/f0c18d4bd76d78eacf4bd4b56643479712089c8a/docker/test/stress/run.sh#L176-L194

Functional: https://github.com/ClickHouse/ClickHouse/blob/f2ef2ec0120374f8beab8e546c09c09ebe620125/tests/clickhouse-test#L537 or https://github.com/ClickHouse/ClickHouse/blob/735f543aeb0ca1fc8dd630bb01e1c11bff994b3f/docker/test/stateless/run.sh#L30-L42

serxa replied:
Done. I added some kind of "randomization". It's not random at all actually, based on the even/odd day of the month. But if you have better ideas, welcome. There is currently no server setting randomization...

serxa commented on src/Storages/System/StorageSystemAsyncLoader.cpp:
```diff
@@ -0,0 +1,142 @@
+            {
+                std::rethrow_exception(state.job->exception());
+            }
+            catch (Exception & e)
+            {
+                exception = e.message();
+            }
+            catch (...) // just in case
+            {
+                exception = String("unknown error");
```
AsyncLoader wraps every exception it gets with `DB::Exception`. But let's fix it. Anyway `getCurrentExceptionMessage` fells cleaner.

serxa commented on src/Storages/System/StorageSystemAsyncLoader.cpp:
```diff
@@ -0,0 +1,142 @@
+        { "is_ready",           std::make_shared<DataTypeUInt8>() },
+        { "elapsed",            std::make_shared<DataTypeFloat64>()},
+        { "pool_id",            std::make_shared<DataTypeInt64>() },
+        { "pool",               std::make_shared<DataTypeString>() },
+        { "priority",           std::make_shared<DataTypeInt64>() },
+        { "execution_pool_id",  std::make_shared<DataTypeInt64>() },
+        { "execution_pool",     std::make_shared<DataTypeString>() },
```
Yes, I'll add

serxa commented on src/Common/AsyncLoader.cpp:
```diff
@@ -571,35 +577,40 @@ void AsyncLoader::finish(const LoadJobPtr & job, LoadStatus status, std::excepti
+    Pool & old_pool = pools[job->pool_id];
+    Pool & new_pool = pools[new_pool_id];
+    if (old_pool.priority <= new_pool.priority)
+        return; // Never lower priority or change pool leaving the same priority
```
I thought it was intended to mark the documentation and to distinguish it from just regular comments for future readers. Some popular tools for generating docs use the same notion, but we are not using these tools. IMO it is not necessary for comments inside functions, but is good for classes and method declarations. I'll fix it in a separate PR

tavplubix replied:
> I'll fix it in a separate PR

It's not necessary, it's okay to keep comments in existing files as is and use `///` in new files

serxa commented on src/Common/AsyncLoader.cpp:
```diff
@@ -619,19 +630,113 @@ void AsyncLoader::enqueue(Info & info, const LoadJobPtr & job, std::unique_lock<
+    if (job->load_status != LoadStatus::PENDING)
+        return; // Job is already done, worker can continue execution
+
+    // To resolve "blocked pool" deadlocks we spawn a new worker for every suspended worker, if required
+    // This can lead to a visible excess of `max_threads` specified for a pool,
+    // but actual number of NOT suspended workers may exceed `max_threads` ONLY in intermittent state.
+    Pool & pool = pools[pool_id];
+    pool.suspended_workers++;
+    job->suspended_waiters++;
+    if (canSpawnWorker(pool, lock))
+        spawn(pool, lock);
```
Correct. It is only limited by the number of "waits" from other jobs. Which is already an unusual operation. Regular dependencies between jobs do not create suspended workers. In this PR we can have it only when while loading a table (from load job) it calls getTable() on the other table that for some reason was not in its dependency list.

This kind of wait will suspend the waiter and increase the pool size by one. This is done to resolve the problem of exhausting the pool with waiting threads, which is a deadlock.

serxa commented on src/Common/AsyncLoader.cpp:
```diff
@@ -619,19 +630,113 @@ void AsyncLoader::enqueue(Info & info, const LoadJobPtr & job, std::unique_lock<
+
+void AsyncLoader::wait(std::unique_lock<std::mutex> & job_lock, const LoadJobPtr & job)
+{
+    // Ensure job we are going to wait was scheduled to avoid "wait not scheduled" deadlocks
+    if (job->job_id == 0)
+    {
+        job_lock.unlock(); // Avoid reverse locking order
+        schedule(LoadJobSet{job});
+        job_lock.lock();
+        chassert(job->job_id != 0);
+    }
```
Yes, let's try. Should work better.

serxa commented on src/Common/AsyncLoader.h:
```diff
@@ -464,4 +392,120 @@ class AsyncLoader : private boost::noncopyable
+    waitLoad(tasks);
+}
+
+inline LoadJobSet getGoals(const LoadTaskPtrs & tasks)
+{
+    LoadJobSet result;
+    for (const auto & task : tasks)
+        result.insert(task->goals().begin(), task->goals().end());
+    return result;
+}
```
Only in tests, in real code `getGoalsOr()` turned out to be more useful. Okay let's get rid of it.

serxa committed:
working on review comments

serxa committed:
rename settings

serxa committed:
add explanations for helper functions

serxa committed:
do CREATE query async

serxa committed:
fix style

serxa committed:
fix spell check

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
create async_loader once

serxa committed:
add startup waits in all DDL queries and global DDL worker

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
get rid of atomics

serxa committed:
fix build

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix conflicts

serxa committed:
implement waits inside tables iterator

serxa committed:
remove wrong waits

serxa committed:
fix

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
add AsyncLoaderWaitMicroseconds profile event

serxa committed:
`DatabaseCatalog::getTable()` should also wait

serxa commented:
One of tests is failed. Looks like there is some race between db start and stop. Or therecan exist databases that are not started for some reason? 
https://pastila.nl/?000f2a02/59fb6e6fd0d2ead8d16f01aa0e646b20#GmKKY0qKf3D8fBppwBTtOw==

serxa committed:
wait for db to be started iff it was async-started

serxa committed:
better

serxa committed:
randomize async_load_databases in stress tests

serxa committed:
fix defaults for new pools in server config.xml

serxa committed:
rename metrics for new pools to match settings

serxa committed:
rework AsyncLoaderPoolId.h to match settings and metrics

serxa committed:
fix ddl worker startup

serxa commented:
`test_delayed_replica_failover` — flaky. Other tests are green, finally!
Now let's trigger stress tests with `async_load_databases = true` to check if that works

serxa committed:
comment

serxa committed:
Merge branch 'master' into async-loader-integration

alexey-milovidov commented:
@serxa, something interesting around `DB::DatabaseAtomic::assertDetachedTableNotInUse`

serxa committed:
Merge branch 'master' into async-loader-integration

serxa committed:
fix

serxa committed:
Merge branch 'master' into async-loader-integration

serxa commented:
> something interesting around DB::DatabaseAtomic::assertDetachedTableNotInUse

I'm still trying to figure out what went wrong. I found a backtrace with variables in gdb.log of the stress test and according to table_id in one of the frames the table in question is `test_ghlfanhj`.`dict2`:
```
(comment: 01160_table_dependencies.sh) CREATE DICTIONARY dict2 (`n` int DEFAULT 0, `m` int DEFAULT 2) PRIMARY KEY n SOURCE(CLICKHOUSE(HOST 'localhost' PORT tcpPort() USER 'default' TABLE 'join' PASSWORD '[HIDDEN]' DB 'test_ghlfanhj')) LIFETIME(MIN 1 MAX 10) LAYOUT(FLAT())
```
This test is very good at checking issues with dependencies. It had already found a deadlock (fixed) in this PR.

serxa commented:
Looks like detach/attach database cycle in `01160_table_dependencies` test lead to some problems and eventually to the crash:
https://pastila.nl/?00008e27/2952cd9a4a402f3e6643a26281147283#4oSjapDCJRZzxZbRtTPBLQ==

serxa committed:
wait for database to load before shutdown

serxa commented:
After fix materialized mysql database became broken:
https://pastila.nl/?0064b1cc/eb259a2b9f6f477c667dae73b144a135#3gTFbPYFyCp0IkNYFY60hQ==

alexey-milovidov commented:
Just in case, MaterializedMySQL is a second-class feature - if needed, you can make a shortcut so that it is loaded non-lazily.

serxa commented:
> you can make a shortcut so that it is loaded non-lazily.

MaterializedMySQL is inherited from AtomicDatabase, which is now reworked to be async only. So it might be even harder than figure out what when wrong... Checking...

serxa commented:
The problem is here:
```python
    clickhouse_node.query(f"ALTER NAMED COLLECTION {db} SET port=9999")
    clickhouse_node.query_with_retry(f"DETACH DATABASE {db} SYNC")
    mysql_node.query(f"INSERT INTO {db}.t1 VALUES (3, 'c', 3)")
    assert "ConnectionFailed:" in clickhouse_node.query_and_get_error(
        f"ATTACH DATABASE {db}"
    )
    clickhouse_node.query(f"ALTER NAMED COLLECTION {db} SET port=3306")
    clickhouse_node.query(f"ATTACH DATABASE {db}")
```
`ATTACH DATABASE {db}` throws an exception from another place (later) because the database load is async. This leads to some trash left in DatabaseCatalog. Fixing...

serxa committed:
do not rethrow load errors during db shutdown

serxa committed:
get rid of defaults to make clang-tidy happy

serxa commented:
CI is green now.
`01104_distributed_numbers_test` - Flaky
`test_postgres_dictionaries_custom_query_full_load` - Leak in PQmakeEmptyPGresult - unrelated

danthegoodman1 commented:
This doesn’t actually speed up start to query time for those tables through right? Just for getting a sql session/health checks?

serxa commented:
> This doesn’t actually speed up start to query time for those tables though right? Just for getting a sql session/health checks?

No. If the server has a lot of tables to load at startup, this feature will also improve the time for queries due to prioritization. Prioritization means loading of tables that are being waited for by queries is done first.

azat commented on src/Interpreters/InterpreterCreateQuery.cpp:
```diff
@@ -325,11 +327,22 @@ BlockIO InterpreterCreateQuery::createDatabase(ASTCreateQuery & create)
-            loader.startupTables();
+            auto load_tasks = loader.loadTablesAsync();
+            auto startup_tasks = loader.startupTablesAsync();
+            if (getContext()->getGlobalContext()->getServerSettings().async_load_databases)
```
@serxa does it make sense to load databases asynchronously in case of ATTACH query? It may only surprise users I guess, and you can scale it by your own anyway (unlike server startup)

serxa replied:
Yes, now any load is asynchronous, we can either wait or not wait.

serxa replied:
In most cases, it is okay to do it async and not even wait, because all operations that require table/database to be loaded will wait for it.

azat replied:
@serxa Do you have objections on making ATTACH queries non-async?

serxa replied:
No, I think we could change this to what it was before. It was just a way to do everything simple. But sync wait looks more like previous behaviour indeed.
